client/core/core.go

```diff
@@ -2586,28 +2586,56 @@ func (c *Core) handleReconnect(host string) {
 		return
 	}
 
-	dc.marketMtx.RLock()
-	for mktKey, mkt := range dc.marketMap {
+	resubMkt := func(mkt *Market) {
+		// Locate any bookie for this market.
 		dc.booksMtx.Lock()
-		booky, foundMkt := dc.books[mktKey]
-		dc.booksMtx.Unlock()
+		defer dc.booksMtx.Unlock() // keep it locked for the whole operation
+		booky, foundMkt := dc.books[mkt.Name]
 		if !foundMkt {
-			continue
+			// Not subscribed with the server for this market.
+			return
 		}
-		feeds := booky.feeds
-		dc.booksMtx.Lock()
-		delete(dc.books, mktKey)
-		dc.booksMtx.Unlock()
-		_, _, err := c.Sync(host, mkt.BaseID, mkt.QuoteID)
+
+		// Unregister the bookie.
+		delete(dc.books, mkt.Name)
+
+		// Deep copy the old BookFeeds map.
+		booky.mtx.Lock()
+		feeds := make(map[uint32]*BookFeed) // likely 1 or 0, use default cap
+		for id, feed := range booky.feeds {
+			feeds[id] = feed
+		}
+		booky.mtx.Unlock()
+
+		// Resubscribe to orderbook with the server, registering a new bookie in
+		// the dc.books map.
+		_, _, err := c.sync(dc, mkt.BaseID, mkt.QuoteID)
 		if err != nil {
 			log.Errorf("error re-subscribing: %v", err)
-			continue
+			return
 		}
-		dc.booksMtx.Lock()
-		dc.books[mktKey].feeds = feeds
```

Comment:
This was the main problem.  The shallow copy of the map was just putting it under two different mutexes, allowing concurrent writes (in this case delete and assign) from the two different `bookie` instances, one for `bookie` instance A in `(*bookie).closeFeed` when a `marketSyncer` stops thus calling `(*bookie).closeFeed`, and one for `bookie` instance B in `(*bookie).feed` when a second connection starts and tries to make a new feed in `Core.Sync`.